### PR TITLE
Correctly process record declaration at a top level

### LINKED ISSUE #77: Declaration in `record` macro is incorrectly interpreted as useless assignment
```cr
record Foo, foo = "foo" # Useless assignment to variable `foo`
```

I guess all assignments used as macro arguments shouldn't be treated as a useless assign because you can't know how it will be used in the macro.

### PR DESCRIPTION
**Pull request by veelenga**
fixes #77

/cc @straight-shoota

### CODE REVIEW

File: spec/ameba/rule/lint/useless_assign_spec.cr [hunk 1]
```diff
@@ -419,6 +419,35 @@ module Ameba::Rule::Lint
...
+      it "reports if assignment is not referenced after the record declaration" do
+        s = Source.new %(
+          foo = 2
+          record Bar, foo = 3
+        ), "source.cr"
+        subject.catch(s).should_not be_valid
+        s.issues.size.should eq 1
+
+        issue = s.issues.first
+        issue.location.to_s.should eq "source.cr:2:11"
```
Comment (straight-shoota) [inline spec/ameba/rule/lint/useless_assign_spec.cr]: The line number is actually correct, but it is hard to follow why this is in 2:11. Why don't you use heredoc instead of a regular string literal? This makes it much easier.

```cr
source = Source.new <<-'SOURCE',
  foo = 2
  record Bar, foo = 3
  SOURCE
  "source.cr"
# ...
issue.location.to_s.should eq "source.cr:1:1"
```

File: src/ameba/ast/visitors/scope_visitor.cr [hunk 2]
```diff
@@ -3,6 +3,21 @@ require "./base_visitor"
 module Ameba::AST
   # AST Visitor that traverses the source and constructs scopes.
   class ScopeVisitor < BaseVisitor
+    SUPER = "super"
+    # List of special top level macros that use assignments
+    # as the macro arguments thus such arguments are not treated as assignments.
```
Comment (straight-shoota) [inline src/ameba/ast/visitors/scope_visitor.cr]: I'd add a note that this is essentially just guessing that such a call (like "record") probably calls a macro.

File: src/ameba/ast/visitors/scope_visitor.cr [hunk 3]
```diff
@@ -3,6 +3,21 @@ require "./base_visitor"
...
+    #
+    # For example:
+    #
+    # ```
+    # bar = 1
```
Comment (straight-shoota) [inline src/ameba/ast/visitors/scope_visitor.cr]: This line is uneccessary for a demonstration. Ditto two below.

File: src/ameba/ast/visitors/scope_visitor.cr [hunk 4]
```diff
@@ -3,6 +3,21 @@ require "./base_visitor"
 module Ameba::AST
   # AST Visitor that traverses the source and constructs scopes.
   class ScopeVisitor < BaseVisitor
+    SUPER = "super"
```
Comment (straight-shoota) [inline src/ameba/ast/visitors/scope_visitor.cr]: What is this constant for? Unlike Ruby (where strings are mutable), Crystal doesnt gain any benefit from extracting string literals to constants.

File: src/ameba/ast/visitors/scope_visitor.cr [hunk 5]
```diff
@@ -140,12 +155,18 @@ module Ameba::AST
...
+      case @current_scope
+      when .def?
+        if node.name == SUPER && node.args.empty?
+          @current_scope.arguments.each do |arg|
+            variable = arg.variable
+            variable.reference(variable.node, @current_scope).explicit = false
+          end
+        end
+      when .top_level?
+        return false if SPECIAL_TOP_LEVEL_MACROS.includes?(node.name)
```
Comment (straight-shoota) [inline src/ameba/ast/visitors/scope_visitor.cr]: For higher precision matching the signature of the `record` macro, you could check if the first argument is a `Crystal::Path`. This obviously requires more customization than just a list of special macro names.

File: spec/ameba/rule/lint/useless_assign_spec.cr (hunk 1 above)
Comment (veelenga) [reply spec/ameba/rule/lint/useless_assign_spec.cr]: This is a convention I follow across the project. Heredoc syntax actually looks weird itself and I don't want to use it here.

File: src/ameba/ast/visitors/scope_visitor.cr (hunk 4 above)
Comment (veelenga) [reply src/ameba/ast/visitors/scope_visitor.cr]: Constant is created once across the application, literals (in a method) are created each time you call a method.

Comment (veelenga) [reply src/ameba/ast/visitors/scope_visitor.cr]: We already discussed that https://github.com/veelenga/ameba/pull/50#issuecomment-381723855

File: spec/ameba/rule/lint/useless_assign_spec.cr (hunk 1 above)
Comment (straight-shoota) [reply spec/ameba/rule/lint/useless_assign_spec.cr]: Well, I think `source.cr:2:11` looks super weird. If you don't want to use heredoc, fine. But maybe you can make this less weird in a different way...

File: src/ameba/ast/visitors/scope_visitor.cr (hunk 3 above)
Comment (straight-shoota) [reply src/ameba/ast/visitors/scope_visitor.cr]: `record Foo, bar = 2 # this is not an assignment` says it all.

File: spec/ameba/rule/lint/useless_assign_spec.cr (hunk 1 above)
Comment (veelenga) [reply spec/ameba/rule/lint/useless_assign_spec.cr]: Okay. This is a topic of a different PR.

File: src/ameba/ast/visitors/scope_visitor.cr (hunk 3 above)
Comment (veelenga) [reply src/ameba/ast/visitors/scope_visitor.cr]: Thanks. I would prefer to leave it as is.
